Testnet/Ticketbooth/Ticketbooth/TicketContract_1_0_0.cs

```diff
@@ -0,0 +1,474 @@
+        }
+
+        Tickets = copyOfTickets;
+        EndOfSale = endOfSale;
+
+        var show = new Show
+        {
+            Name = showName,
+            Organiser = organiser,
+            Time = time,
+            EndOfSale = endOfSale
+        };
+        Log(show);
+    }
+
+    /// <summary>
+    /// Called after the ending of a ticket sale to clear the contract ticket data
+    /// </summary>
+    public void EndSale()
+    {
+        Assert(Message.Sender == Owner, "Only contract owner can end sale");
+        Assert(EndOfSale != default(ulong), "Sale not currently in progress");
+        Assert(Block.Number >= EndOfSale, "Sale contract not fulfilled");
+
+        Tickets = ResetTickets(Tickets);
+        EndOfSale = default(ulong);
+    }
+
+    /// <summary>
+    /// Checks the availability of a seat
+    /// </summary>
+    /// <param name="seatIdentifierBytes">The serialized seat identifier</param>
+    /// <returns>Whether the seat is available</returns>
+    public bool CheckAvailability(byte[] seatIdentifierBytes)
+    {
+        Assert(SaleOpen, "Sale not open");
+
+        var ticket = SelectTicket(seatIdentifierBytes);
+
+        Assert(!IsDefaultSeat(ticket.Seat), "Seat not found");
+
+        return IsAvailable(ticket);
+    }
+
+    /// <summary>
+    /// Reserves a ticket for the callers address
+    /// </summary>
+    /// <param name="seatIdentifierBytes">The serialized seat identifier</param>
+    /// <param name="secret">The encrypted secret holding ticket ownership</param>
+    /// <returns>Whether the seat was successfully reserved</returns>
+    public void Reserve(byte[] seatIdentifierBytes, byte[] secret)
+    {
+        Reserve(seatIdentifierBytes, secret, null);
+    }
+
+    /// <summary>
+    /// Reserves a ticket for the callers address and with an identifier for the customer
+    /// </summary>
+    /// <param name="seatIdentifierBytes">The serialized seat identifier</param>
+    /// <param name="secret">The encrypted secret holding ticket ownership</param>
+    /// <param name="customerIdentifier">An encrypted verifiable identifier for the customer</param>
+    /// <returns>Whether the seat was successfully reserved</returns>
+    public void Reserve(byte[] seatIdentifierBytes, byte[] secret, byte[] customerIdentifier)
+    {
+        Assert(secret != null, "Invalid secret");
+        Assert(!RequireIdentityVerification || customerIdentifier != null, "Invalid customer identifier");
+        Assert(SaleOpen, "Sale not open");
+
+        var seat = Serializer.ToStruct<Seat>(seatIdentifierBytes);
+        var copyOfTickets = Tickets;
+        Ticket ticket = default(Ticket);
+        int ticketIndex = 0;
+        for (var i = 0; i < copyOfTickets.Length; i++)
+        {
+            if (SeatsAreEqual(copyOfTickets[i].Seat, seat))
+            {
+                ticketIndex = i;
+                ticket = copyOfTickets[i];
+                break;
+            }
+        }
```

Comment:
Anyway to avoid loops here in some of these implementations? Maybe a local call from UI to get the tickets then finding the index outside of the contract? My only concern here is that the gas costs might get high depending on the number of tickets. Selling 50 tickets may be no big deal, selling 10,000 tickets might be.

Reply:
That's an interesting idea. You're right, having these loops means that the contract can only handle around 70 tickets with the gas limit of 100000. Doing what you suggested could possibly increase this by ~15 to 20%, as starting the sale has a higher cost than creating the contract. There isn't much room, if any, to reduce the cost of starting a sale.

I've thought about this and am planning to create a v2 which will include a 'management' contract that can create sale contracts. Each individual contract would be for a group of seats and this will allow it to scale up to 10000+ seats.

For now I'm inclined to stick with this implementation, but I could test out your idea and implement it in a future version. It definitely would save gas.

Reply:
Fair enough, might be a good idea then to limit the amount of tickets that can be created during the creation of the smart contract. The limit being the maximum amount that can be supported for any loop in any method. Would need to also leave room here because users could adjust the gas cost to be higher than the default 100.